File: solutions/javascript/dfs/wordSearch.js
const directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
];

const getPositionString = (row, col) => `${row}, ${col}`;

const isInBounds = (board, row, col) => {
    const numRows = board.length;
    const numCols = board[0].length;

    const rowInBounds = row >= 0 && row < numRows;
    const colInBounds = col >= 0 && col < numCols;

    return rowInBounds && colInBounds;
};

const checkIfWordExists = (board, row, col, visited, word, wordIdx) => {
    // Base cases
    const curPositionString = getPositionString(row, col);

    if (wordIdx >= word.length) return true;

    if (!isInBounds(board, row, col)) return false;
    if (visited.has(curPositionString)) return false;

    // Process node
    const boardChar = board[row][col];
    const wordChar = word[wordIdx];

    visited.add(curPositionString);

    if (boardChar !== wordChar) {
        visited.delete(curPositionString);
        return false;
    }

    // Recurse on neighbors
    for (const direction of directions) {
        const [rowChange, colChange] = direction;

        const newRow = row + rowChange;
        const newCol = col + colChange;

        const remainingWordExists = checkIfWordExists(
            board,
            newRow,
            newCol,
            visited,
            word,
            wordIdx + 1,
        );
        if (remainingWordExists) {
            visited.delete(curPositionString);
            return true;
        }
    }

    visited.delete(curPositionString);
    return false;
};

const exist = (board, word) => {
    const numRows = board.length;
    const numCols = board[0].length;

    // const wordFirstChar = word[0];
    const visited = new Set();

    for (let row = 0; row < numRows; row += 1) {
        for (let col = 0; col < numCols; col += 1) {
            const wordExistsHere = checkIfWordExists(board, row, col, visited, word, 0);
            if (wordExistsHere) return true;
        }
    }

    return false;
};

File: solutions/python/dfs/word_search.py
directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
]

def is_in_bounds(board, row, col):
    num_rows = len(board)
    num_cols = len(board[0])

    row_in_bounds = row >= 0 and row < num_rows
    col_in_bounds = col >= 0 and col < num_cols

    return row_in_bounds and col_in_bounds

class Solution:
    def check_if_word_exists(self, board, row, col, visited, word, word_idx):
        # Base cases
        if word_idx >= len(word):
            return True

        if not is_in_bounds(board, row, col):
            return False

        if (row, col) in visited:
            return False

        # Process node
        board_char = board[row][col]
        word_char = word[word_idx]

        visited.add((row, col))

        if board_char != word_char:
            visited.remove((row, col))
            return False

        # Recurse on neighbors
        for direction in directions:
            row_change, col_change = direction

            new_row = row + row_change
            new_col = col + col_change

            remaining_word_exists = self.check_if_word_exists(
                board,
                new_row,
                new_col,
                visited,
                word,
                word_idx + 1,
            )

            if remaining_word_exists:
                visited.remove((row, col))
                return True
            
        visited.remove((row, col))
        return False

    def exist(self, board: List[List[str]], word: str) -> bool:
        num_rows = len(board)
        num_cols = len(board[0])

        visited = set()

        for row in range(num_rows):
            for col in range(num_cols):
                word_exists_here = self.check_if_word_exists(board, row, col, visited, word, 0)
                if word_exists_here:
                    return True

        return False